### PR TITLE
#14243 AsyncResponse to avoid timeout when restoring big snapshots.

### PR DESCRIPTION
**Pull request by dsilvam**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 1]
```diff
@@ -302,18 +309,15 @@ public Response snapshotIndex(@Context HttpServletRequest request, @PathParam("p
...
-			};
+			StreamingOutput stream = os -> {
+                try {
+                    ByteStreams.copy(in, os);
+                } finally {
+                    // clean up
+                    indexAPI.deleteRepository(ESIndexAPI.BACKUP_REPOSITORY, true);
+                    snapshotFile.delete();
+                }
+            };
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: this is just replacing the anonymous class with lambda

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 2]
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
...
         }catch(IllegalArgumentException iar){
         	return this.responseUtil.getErrorResponse(request, Response.Status.BAD_REQUEST, Locale.getDefault(), null, "snapshot.wrong.arguments");
-        }catch(IOException ex) {
-        	return this.responseUtil.getErrorResponse(request, Response.Status.INTERNAL_SERVER_ERROR, Locale.getDefault(), null, "snapshot.io.writing.fail");
-        }catch(ExecutionException exc){
-        	return this.responseUtil.getErrorResponse(request, Response.Status.INTERNAL_SERVER_ERROR, Locale.getDefault(), null, "snapshot.fail.during.execution");
-        }catch(SnapshotRestoreException exc){
-        	return this.responseUtil.getErrorResponse(request, Response.Status.INTERNAL_SERVER_ERROR, Locale.getDefault(), null, "snapshot.fail");
-        }catch(InterruptedException exi){
-        	return this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: non of these are thrown anymore

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 3]
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
...
+                Response response = null;
+                try {
+                    if(this.indexAPI.uploadSnapshot(inputFile)){
+                        response = Response.ok(new MessageEntity("Snapshot restored successfully")).build();
+                    }
+                } catch (InterruptedException | ExecutionException | IOException e) {
+                    response = this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
+                }
+
+                asyncResponse.resume(response);
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: we resume the response after we are done (good or bad) with the restoring

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 4]
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
...
-            }
+        	auth(params,request);
+
+            DotSubmitter submitter = DotConcurrentFactory.getInstance().getSubmitter(DOT_SYSTEM_THREAD_POOL);
+
+            submitter.submit(() -> {
+                Response response = null;
+                try {
+                    if(this.indexAPI.uploadSnapshot(inputFile)){
+                        response = Response.ok(new MessageEntity("Snapshot restored successfully")).build();
```
Comment (acurionedotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: Can this be externalized ?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 5]
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
...
+                    }
+                } catch (InterruptedException | ExecutionException | IOException e) {
+                    response = this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: Is it worth to log this exception?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java [hunk 6]
```diff
@@ -343,31 +347,36 @@ public void write(OutputStream os) throws IOException, WebApplicationException {
...
+                }
+
+                asyncResponse.resume(response);
+            });
             return this.responseUtil.getErrorResponse(request, Response.Status.SERVICE_UNAVAILABLE, Locale.getDefault(), null, "snapshot.upload.failed");
 
         }catch (SecurityException sec) {
             return this.responseUtil.getErrorResponse(request, Response.Status.UNAUTHORIZED, Locale.getDefault(), null, "authentication-failed");
         }catch(IllegalArgumentException iar){
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: same comment here

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: I think so. Will add it

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java (hunk 6 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java]: will add them also

**Review by jdotcms — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Review by fabrizzio-dotCMS — APPROVED** [review]

**Review by acurionedotcms — APPROVED** [review]

**Pull request merged**